CI: validate build on Win-ARM64

MugundanMCW opened the pull request:
<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
https://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
https://scipy.github.io/devdocs/dev/contributor/development_workflow.html#writing-the-commit-message.
However, please only include an issue number in the description, not the title,
and please ensure that any code names containing underscores are enclosed in backticks.

Depending on your changes, you can skip CI operations and save time and energy: 
https://scipy.github.io/devdocs/dev/contributor/continuous_integration.html#skipping

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

#### Reference issue
<!--Example: Closes gh-WXYZ.-->
Closes: https://github.com/scipy/scipy/issues/21562 https://github.com/scipy/scipy/issues/22825

#### What does this implement/fix?
<!--Please explain your changes.-->
Enables new CI environment for building & validating SciPy for Win-ARM64.

#### Additional information
<!--Any additional information you think is important.-->

- The PR introduces setting up a new CI environment for building and testing SciPy locally for Win-ARM64.
- This CI setup depends on the latest stable LLVM toolchain & vcpkg(for native pkg-config).

rgommers reviewed (commented):
Thanks for working on this @Mugundanmcw. It isn't yet mergeable until we resolve (or skip if needed) the failing tests, but let me ask for a restructure of the CI setup now so things are as maintainable as possible. This is independent of looking into the failures.

- We want these new jobs to live in `windows.yml`, rather than having the in a separate workflow file
- Let's go for 2 jobs rather than 3, that's a bit much.  We want to keep `fast_spin` and `full_build_sdist_wheel`
- Then let's also refactor out the verbose bits related to LLVM and pkg-config. Can you put those in a composite action, `.github/windows_arm64_steps/action.yml`? Just like NumPy does with [.github/meson_actions/action.yml](https://github.com/numpy/numpy/blob/main/.github/meson_actions/action.yml). Then you can call that with `uses:`, just like an external action.

This will make the jobs a lot more concise and easy to understand.

rgommers commented on .github/workflows/windows.yml:
```diff
@@ -168,3 +168,106 @@ jobs:
+    runs-on: windows-11-arm
+    # if: > 
+    #   needs.get_commit_message.outputs.message == 1 &&
+    #   (github.repository == 'scipy/scipy' || github.repository == '')
```
This needs to be uncommented.

rgommers commented on .github/workflows/windows.yml:
```diff
@@ -168,3 +168,106 @@ jobs:
+          python-version: '3.12'
+          architecture: 'arm64'
+          
+      - name: Common Setup
```
How about "Set up Flang and pkgconf for arm64" as a descriptive name?

Mugundanmcw replied:
Look precise, will change it accordingly

rgommers commented on .github/workflows/windows.yml:
```diff
@@ -168,3 +168,106 @@ jobs:
+
+      - name: pip-packages
+        run: |
+          pip install git+https://github.com/mesonbuild/meson.git@ddc01cf5d9240cdb0edc4b44e837260948441d12
```
This can be dropped now that 1.8.2 is out

rgommers commented on .github/workflows/windows.yml:
```diff
@@ -168,3 +168,106 @@ jobs:
+    needs: get_commit_message
+    # if: >
+    #   needs.get_commit_message.outputs.message == 1 &&
+    #   (github.repository == 'scipy/scipy' || github.repository == '')
```
can be uncommented

rgommers commented on .github/workflows/windows.yml:
```diff
@@ -168,3 +168,106 @@ jobs:
+  
+      - name: pip-packages
+        run: |
+          pip install git+https://github.com/mesonbuild/meson.git@ddc01cf5d9240cdb0edc4b44e837260948441d12
```
can be dropped, move `meson` to the next line

rgommers commented on .github/workflows/windows.yml:
```diff
@@ -168,3 +168,106 @@ jobs:
+        run: |
+          Set-Location $env:RUNNER_TEMP
+          $env:OPENBLAS_NUM_THREADS = 1
+          pytest --pyargs scipy
```
I suggest adding `-n2` for parallel testing with `pytest-xdist`. It'd also be nice to check if `-n3` or `-n4` is even faster.

Mugundanmcw replied:
I kept n2 for parallel testing. I can probably trigger my fork CI with n3/n4 and check whether it provides better execution.

rgommers commented on .github/workflows/windows.yml:
```diff
@@ -168,3 +168,106 @@ jobs:
+          python -m build --no-isolation -x -Csetup-args="-Duse-pythran=false"
+          wheel_name=$(ls dist/*.whl)
+          openblas_dir=$(python -c"import scipy_openblas32 as sop; print(sop.get_lib_dir())")
+          delvewheel repair --add-path "$openblas_dir" --no-dll libsf_error_state.dll -w wheelhouse "$wheel_name"
```
`libsf_error_state.dll` is gone, so that part shouldn't need to be there anymore.

rgommers commented on .github/workflows/windows.yml:
```diff
@@ -168,3 +168,106 @@ jobs:
+          cache: 'pip'
+          cache-dependency-path: 'environment.yml'
+  
+      - name: Common Setup
```
same suggestion about naming as for the other job

rgommers reviewed (commented):
This is getting closer. A few more suggestions. The `test_lapack` failures are still there unfortunately.

rgommers commented on requirements/openblas.txt:
```diff
@@ -1 +1 @@
-scipy-openblas32==0.3.28.0.2
+scipy-openblas32==0.3.29.265.1
```
This worked and the bump has been made in `main` now, so you can drop it from this PR.

ilayn commented:
> The `test_lapack` failures are still there unfortunately

Yeah, it looks like we are not seeing something obvious that only hits this platform, but not sure what it is yet.

MugundanMCW committed:
Add WOA support to windows.yml

Adding windows on ARM support for windows.yml file to build arm64 wheels
create action.yml to split LLVM and pkgconf changes as reusable blocks.

Mugundanmcw commented:
> `test_lapack` fails only on GHA or locally too?

I tried running the test suite locally on my native WoA device, and I could that the same number of test cases were failing locally too:
![image](https://github.com/user-attachments/assets/5a08939c-3769-4344-830f-e50dfed9150b)

ilayn commented:
This does not seem like a SciPy issue but the underlying LAPACK vendor might have a bug. I tried many inputs with various size but did not manage to crash it.

ilayn commented:
If there is a way to test the OpenBLAS functionality with a simple C/C++/Fortran program by linking to it would clarify things better on this platform. Regular LAPACK repo is the reference implementation that we are not using here. 

There might be an issue on the ARM64 target for sytri/hetri. See for example this one https://github.com/OpenMathLib/OpenBLAS/issues/5293

khmyznikov commented:
@ilayn do you use a fork of LAPACK? How it was built for WoA?

lucascolley commented:
LAPACK is included in OpenBLAS, so I think you are looking for https://github.com/MacPython/openblas-libs/blob/main/tools/build_steps_win_arm64.bat

khmyznikov commented:
@lucascolley what if build lapack separetely via llvm and try to use it in tests? Worth to try?

khmyznikov commented:
@rgommers that's sounds reasonable. SciPy should unblock a lot of other packages even in current state. We can work on fix for the next release.

khmyznikov commented:
@rgommers yeah looks like the only option is to keep APL as external dependency. I guess that's separate topic.

MugundanMCW committed:
CI: Add checksum to verify LLVM Package

dschmitz89 commented:
Now we have a flood of new test failure coming from the ARPACK rewrite.

ilayn commented:
Possibly my fault as a bug since I am also hunting for a complex bug too. The real question is how to reproduce these failures since they all pass in other platforms, but I'll check.

Mugundanmcw commented:
- I am currently investigating LAPACK test failures on Windows on ARM (WoA) devices using both MSVC and LLVM toolchains.
- When building SciPy from the main branch on a native Windows on Arm (WoA) device using either MSVC or LLVM’s clang-cl with the flang-new compiler:
  - All test_lapack cases passed in both MSVC and LLVM build configurations,
  - All test_arpack cases failed on both MSVC and LLVM build configurations.
  - Additionally, the test_linprog(test_bug_6139) fails only with MSVC build configuration.
- The new arpack test failures appear to be related to the recent [migration of ARPACK](https://github.com/scipy/scipy/commit/14dffda0b83c46728a7c88998804a4663bb735b3) from Fortran to C. Building SciPy from the [commit](https://github.com/scipy/scipy/commit/c627bcc219b2ebd5117eef352f969cdb4d038a0d) just prior to the ARPACK C migration results in all tests passing with the LLVM toolchain. However, scipy.linprog test failures still persist with the MSVC toolchain in this configuration.
- Triggering CI on the latest main branch with either the MSVC or LLVM configurations, the LAPACK tests still fail due to an access violation exception — an issue that appears to occur only on WoA GitHub Runners, and not locally on native WoA devices.

| SciPy Branch                         | Number of failures          | Failed Test Cases - MSVC               | Failed Test Cases - LLVM  |
|-------------------------------------|-----------------------------|----------------------------------------|---------------------------|
| [Main](https://github.com/scipy/scipy/commit/da62128ca614776e694098590b7c590b48f0933c)         | MSVC: 146<br>LLVM: 144      | `test_arpack(144)`, `test_linprog(2)` | `test_arpack(144)`        |
| [Commit before ARPACK C migration](https://github.com/scipy/scipy/commit/c627bcc219b2ebd5117eef352f969cdb4d038a0d) | MSVC: 2<br>LLVM: 0         | `test_linprog(2)`                      | 0                         |

rgommers commented:
The `test_arpack` failures aren't specific to `arm64`, I'm seeing 144 new failures on Windows x86-64 when building with `-Duse-g77-abi`, all on `test_hermitian` and `test_complex_nonsymmetric_modes`. The numerical results are way off; that is very likely due to the ARPACK rewrite not integrating the BLAS plumbing correctly. I'll open a new issue.

lucascolley commented:
on my windows box I also see

```console
scipy\sparse\linalg\_eigen\arpack\tests\test_arpack.py ............................................................................................. [ 45%]
.................................................................................................................................................... [ 45%]
F.F.F.F.F.F.F.F.F.F.F.F.F.F.F.F.F.F.F.F.F.F.F.F.F.F.F.F.F.F.F.F.F.F.F.F.F.F.F.F.F.F.F.F.F.F.F.F.F.F.F.F.F.F.F.F.F.F.F.F.F.F.F.F.F.F.F.F.F.F.F.F..... [ 45%]
.................................................................................................................................................... [ 46%]
.............FF.FF.FF.FF.FF.FF.FF.FF.FF.FF.FF.FF.FF.FF.FF.FF.FF.FF.FF.FF.FF.FF.FF.FF.FF.FF.FF.FF.FF.FF.FF.FF.FF.FF.FF.FF...........................  [ 46%]
```

lucascolley commented:
that doesn't compile for me:

```console
[3/6] Compiling C object scipy/sparse/linalg/_eigen/arpack/libarnaud.a.p/arnaud_src_arnaud_n_single_complex.c.obj
FAILED: [code=1] scipy/sparse/linalg/_eigen/arpack/libarnaud.a.p/arnaud_src_arnaud_n_single_complex.c.obj
"clang-cl" "-Iscipy\sparse\linalg\_eigen\arpack\libarnaud.a.p" "-Iscipy\sparse\linalg\_eigen\arpack" "-I..\scipy\sparse\linalg\_eigen\arpack" "-I..\scipy\sparse\linalg\_eigen\arpack\arnaud\include" "-IE:/dev/pixi-dev-scipystack/scipy/.pixi/envs/default/Library/include/openblas" "/MD" "/nologo" "/showIncludes" "/utf-8" "/W2" "/clang:-std=c17" "/O2" "/Zi" "-Wno-unused-but-set-variable" "-Wno-unused-function" "-Wno-conversion" "-Wno-misleading-indentation" "/Fdscipy\sparse\linalg\_eigen\arpack\libarnaud.a.p\arnaud_src_arnaud_n_single_complex.c.pdb" /Foscipy/sparse/linalg/_eigen/arpack/libarnaud.a.p/arnaud_src_arnaud_n_single_complex.c.obj "/c" ../scipy/sparse/linalg/_eigen/arpack/arnaud/src/arnaud_n_single_complex.c
../scipy/sparse/linalg/_eigen/arpack/arnaud/src/arnaud_n_single_complex.c(368,33): error: passing 'ARNAUD_CPLXF_TYPE' (aka 'struct _C_float_complex') to parameter of incompatible type '_Dcomplex' (aka 'struct _C_double_complex')
  368 |             temp = _Cmulcr(conj(workl[iheig + k]), 1.0 / mag_sq);
      |                                 ^~~~~~~~~~~~~~~~
C:\Program Files (x86)\Windows Kits\10\include\10.0.22621.0\ucrt\complex.h(86,48): note: passing argument to parameter '_Z' here
   86 | _ACRTIMP _Dcomplex __cdecl conj(_In_ _Dcomplex _Z);
      |                                                ^
../scipy/sparse/linalg/_eigen/arpack/arnaud/src/arnaud_n_single_complex.c(391,33): error: passing 'ARNAUD_CPLXF_TYPE' (aka 'struct _C_float_complex') to parameter of incompatible type '_Dcomplex' (aka 'struct _C_double_complex')
  391 |             temp = _Cmulcr(conj(workl[iheig + k]), 1.0 / mag_sq);
      |                                 ^~~~~~~~~~~~~~~~
C:\Program Files (x86)\Windows Kits\10\include\10.0.22621.0\ucrt\complex.h(86,48): note: passing argument to parameter '_Z' here
   86 | _ACRTIMP _Dcomplex __cdecl conj(_In_ _Dcomplex _Z);
      |                                                ^
../scipy/sparse/linalg/_eigen/arpack/arnaud/src/arnaud_n_single_complex.c(421,37): error: passing 'ARNAUD_CPLXF_TYPE' (aka 'struct _C_float_complex') to parameter of incompatible type '_Dcomplex' (aka 'struct _C_double_complex')
  421 |                 temp = _Cmulcr(conj(workl[iheig + j]), 1.0 / mag_sq);
      |                                     ^~~~~~~~~~~~~~~~
C:\Program Files (x86)\Windows Kits\10\include\10.0.22621.0\ucrt\complex.h(86,48): note: passing argument to parameter '_Z' here
   86 | _ACRTIMP _Dcomplex __cdecl conj(_In_ _Dcomplex _Z);
      |                                                ^
../scipy/sparse/linalg/_eigen/arpack/arnaud/src/arnaud_n_single_complex.c(1319,23): warning: unused variable 'tmp_cplx2' [-Wunused-variable]
 1319 |     ARNAUD_CPLXF_TYPE tmp_cplx2;
      |                       ^~~~~~~~~
../scipy/sparse/linalg/_eigen/arpack/arnaud/src/arnaud_n_single_complex.c(1519,33): warning: suggest braces around initialization of subobject [-Wmissing-braces]
 1519 |     ARNAUD_CPLXF_TYPE vl[1] = { 0.0 };
      |                                 ^~~
      |                                 {  }
2 warnings and 3 errors generated.
[4/6] Compiling C object scipy/sparse/linalg/_eigen/arpack/libarnaud.a.p/arnaud_src_arnaud_n_double_complex.c.obj
../scipy/sparse/linalg/_eigen/arpack/arnaud/src/arnaud_n_double_complex.c(1319,22): warning: unused variable 'tmp_cplx2' [-Wunused-variable]
 1319 |     ARNAUD_CPLX_TYPE tmp_cplx2;
      |                      ^~~~~~~~~
../scipy/sparse/linalg/_eigen/arpack/arnaud/src/arnaud_n_double_complex.c(1519,32): warning: suggest braces around initialization of subobject [-Wmissing-braces]
 1519 |     ARNAUD_CPLX_TYPE vl[1] = { 0.0 };
      |                                ^~~
      |                                {  }
2 warnings generated.
```

Mugundanmcw commented:
I tested the SciPy main branch with this patch on Windows on ARM device. Upon running test suite, all test_arpack cases were passing with both MSVC and LLVM toolchain. But there is a regression in rotation test case with 1 test failure:

<img width="866" height="141" alt="image" src="https://github.com/user-attachments/assets/fbfc32d5-ca62-443c-b9e2-19374c5d9d06" />

Rotation test failure:
<img width="1397" height="772" alt="image" src="https://github.com/user-attachments/assets/31e1b732-d4cf-482e-92cf-44a979d1e733" />

lucascolley commented:
@Mugundanmcw please try cleaning your environment, this same thing was reported in gh-23302.

Let me make a PR for the patch now.

Mugundanmcw commented:
> @Mugundanmcw please try cleaning your environment, this same thing was reported in [gh-23302](https://github.com/scipy/scipy/issues/23302).

@lucascolley thanks for the suggestion, currently there is no test failures with LLVM for SciPy on WoA device:
<img width="1260" height="522" alt="image" src="https://github.com/user-attachments/assets/c26e3680-b714-4ee7-8102-94e728b88c3d" />

But there are two test failures with MSVC toolchain for SciPy on WoA device:
<img width="1582" height="348" alt="image" src="https://github.com/user-attachments/assets/3eff4da4-fcde-46c1-86d8-d403cb56f62d" />

rgommers commented:
The fix for the ARPACK failures is merged now, so hopefully we're getting close to the end here 🤞🏼

khmyznikov commented:
@rgommers Great! What comes next?

Mugundanmcw commented:
> we still have the failures from [#23133 (comment)](https://github.com/scipy/scipy/pull/23133#issuecomment-2975174582) and `FAILED spatial/tests/test_distance.py::TestCdist::test_cdist_out[mahalanobis]` observed in CI

@rgommers & @lucascolley still running SciPy test suite on Window on ARM locally, I don't see any failures with LLVM and with MSVC same number of test failures (linprog(2 tests)) are observed:

<img width="1555" height="582" alt="Screenshot 2025-07-15 095708" src="https://github.com/user-attachments/assets/cff99f66-4493-4cf9-951b-927956259a40" />

rgommers committed:
STY: strip trailing whitespace

rgommers committed:
TST: skip lapack and spatial.distance.cdist tests that crash on win-arm64

These are both caused by a crash in `np.linalg.env`, see numpy#29442.
The `scipy` functionality under test is fine, verified by separating out
the `linalg.inv` call from the scipy function calls in the problematic
CI job.

rgommers commented:
The crashes are being caused by `numpy.linalg.inv`, see https://github.com/numpy/numpy/issues/29442. That doesn't look good, so if one of you with a WoA dev setup could investigate that would be great.

It's not a blocker for SciPy, so I'll push a couple of commits to this PR to move this along.

rgommers committed:
CI: fix missing pinning of actions and a `needs: get_commit_message`

rgommers reviewed (approved):
The one CI issue is unrelated; this is in good shape now so in it goes. Thanks @Mugundanmcw and everyone else who contributed!

I'll comment on the PR for `win_arm64` wheels next.
